great_expectations/data_context/data_context/abstract_data_context.py

```diff
@@ -58,25 +134,578 @@ def __init__(self, runtime_environment: dict):
+            if AbstractDataContext._default_profilers_exist(
+                directory_path=self.root_directory
+            ):
+                return DataContextConfigDefaults.DEFAULT_PROFILER_STORE_NAME.value
+            if self.root_directory:
+                error_message: str = f'Attempted to access the "profiler_store_name" field with no `profilers` directory.\n  Please create the following directory: {os.path.join(self.root_directory, DataContextConfigDefaults.DEFAULT_PROFILER_STORE_BASE_DIRECTORY_RELATIVE_NAME.value)}\n  To use the new "Profiler Store" feature, please update your configuration to the new version number {float(CURRENT_GE_CONFIG_VERSION)}.\n  Visit https://docs.greatexpectations.io/docs/guides/miscellaneous/migration_guide#migrating-to-the-batch-request-v3-api to learn more about the upgrade process.'
+            else:
+                error_message: str = f'Attempted to access the "profiler_store_name" field with no `profilers` directory.\n  Please create a `profilers` directory in your Great Expectations project " f"directory.\n  To use the new "Profiler Store" feature, please update your configuration to the new version number {float(CURRENT_GE_CONFIG_VERSION)}.\n  Visit https://docs.greatexpectations.io/docs/guides/miscellaneous/migration_guide#migrating-to-the-batch-request-v3-api to learn more about the upgrade process.'
+            raise ge_exceptions.InvalidTopLevelConfigKeyError(error_message)
+
+    @property
+    def profiler_store(self) -> ProfilerStore:
+        profiler_store_name: str = self.profiler_store_name
+        try:
+            return self.stores[profiler_store_name]
+        except KeyError:
+            if AbstractDataContext._default_profilers_exist(
+                directory_path=self.root_directory
+            ):
+                logger.warning(
+                    f'Profiler store named "{profiler_store_name}" is not a configured store, so will try to use default Profiler store.\n  Please update your configuration to the new version number {float(CURRENT_GE_CONFIG_VERSION)} in order to use the new "Profiler Store" feature.\n  Visit https://docs.greatexpectations.io/docs/guides/miscellaneous/migration_guide#migrating-to-the-batch-request-v3-api to learn more about the upgrade process.'
+                )
+                built_store: Optional[Store] = self._build_store_from_config(
+                    profiler_store_name,
+                    DataContextConfigDefaults.DEFAULT_STORES.value[profiler_store_name],
+                )
+                return cast(ProfilerStore, built_store)
+
+            raise ge_exceptions.StoreConfigurationError(
+                f'Attempted to access the Profiler store named "{profiler_store_name}", which is not a configured store.'
+            )
+
+    @property
+    def concurrency(self) -> Optional[ConcurrencyConfig]:
+        return self.project_config_with_variables_substituted.concurrency
+
+    def add_datasource(
```

Comment:
I think the `usage_statistics_enabled_method` decorator should be here in the parent class. It would be worth checking to see if there other instances in this PR of methods where the `usage_statistics_enabled_method` decorator is not in the parent class but only in the child class.

Reply:
there are definitely going to be a few of those.. I tried to keep all the decorators in the `BaseDataContext` class for now, and was thinking of tackling the `usage_stats` decorators in a follow-up PR. 

What are your thoughts on the sequencing ? Should it be handled this time around?

Reply:
My concern was if someone uses these new classes that the usage stats would not be tracked. If it is a fast follow and before these classes are documented for general public use I'm ok with leaving it to a follow-on PR.

Reply:
From discussion : We will need to create new events for each "type" of DataContext. This task is being tracked by GREAT-961